meerk40t/gui/toolwidgets/toolpointlistbuilder.py

```diff
@@ -0,0 +1,238 @@
+                and abs(
+                    complex(*self.point_series[-2]) - complex(*self.point_series[-1])
+                )
+                < 5000
+            ):
+                self.point_series.pop(-1)
+                self.end_tool()
+                response = RESPONSE_ABORT
+        elif event_type == "rightdown":
+            was_already_empty = len(self.point_series) == 0
+            self.end_tool()
+            if was_already_empty:
+                self.scene.context("tool none\n")
+            response = RESPONSE_CONSUME
+        elif event_type == "leftdown":
+            self.scene.pane.tool_active = True
+            if nearest_snap is None:
+                self.mouse_position = space_pos[0], space_pos[1]
+            else:
+                self.mouse_position = nearest_snap[0], nearest_snap[1]
+            if self.point_series:
+                self.scene.request_refresh()
+            response = RESPONSE_CONSUME
+        elif event_type in ("leftup", "move", "hover"):
+            if nearest_snap is None:
+                self.mouse_position = space_pos[0], space_pos[1]
+            else:
+                self.mouse_position = nearest_snap[0], nearest_snap[1]
+            if self.point_series:
+                self.scene.request_refresh()
+                response = RESPONSE_CONSUME
+        elif event_type == "doubleclick":
+            self.end_tool()
+            response = RESPONSE_ABORT
+        elif event_type == "lost" or (event_type == "key_up" and modifiers == "escape"):
+            if self.scene.pane.tool_active:
+                self.scene.pane.tool_active = False
+                self.scene.request_refresh()
+                response = RESPONSE_CONSUME
+            else:
+                response = RESPONSE_CHAIN
+            self.point_series = []
+            self.mouse_position = None
+            self.scene.context.signal("statusmsg", "")
+            self.aborted()
+        elif event_type == "key_up":
+            if not self.scene.pane.tool_active:
+                return RESPONSE_CHAIN
+            if self.key_up(keycode, modifiers):
+                response = RESPONSE_CONSUME
+        elif update_required:
+            self.scene.request_refresh()
+            response = RESPONSE_CONSUME
+        return response
+
+    def aborted(self):
+        # Can be overloaded
+        return
+
+    def key_up(self, keycode, modifiers):
+        # No action, so pass along the code...
+        return False
+
+    def point_added(self):
+        # Can be overloaded
+        # self.point_series does contain the list of valid points
+        return
+
+    def end_tool(self):
+        followup = self.create_node()
+        self.scene.pane.tool_active = False
+        self.point_series = []
+        self.mouse_position = None
+        self.scene.request_refresh()
+        if followup:
+            self.scene.context(f"{followup}\n")
+
+    def create_node(self):
+        print("Needs to be overloaded!")
```

Comment:
This should `raise NotImplementedError`

exception NotImplementedError[¶](https://docs.python.org/3/library/exceptions.html#NotImplementedError)
This exception is derived from [RuntimeError](https://docs.python.org/3/library/exceptions.html#RuntimeError). In user defined base classes, abstract methods should raise this exception when they require derived classes to override the method, or while the class is being developed to indicate that the real implementation still needs to be added.

Reply:
Okay, makes sense